Elasticsearch

pkarman opened the pull request:
Basic infrastructure for building ES indices via rake task. One API endpoint partially implemented for Story model.

Still TODO via Epic #444
* [ ] Re-implement API filters against ES index for Story #441
* [ ] Re-implement Networks search for non-Story models #442
* [ ] Incremental updates #443

cavis commented on app/models/concerns/searchable.rb:
```diff
@@ -0,0 +1,72 @@
+      self.class.__elasticsearch__.refresh_index!
+    end
+
+    def self.rebuild_index
```
What is this used for?  (Not called anywhere in this PR).  Is this the high-availability async reindexing?

cavis commented on app/services/story_query_builder.rb:
```diff
@@ -0,0 +1,192 @@
+    end
+  end
+
+  def munge_fielded_params(fielded)
```
What is this used for?  Some sort of created-in-range query that we don't use yet?

pkarman replied:
yes. I basically lifted a query builder class I built a couple years ago and started adapting. There are several vestigial pieces. I will mark those so it's clear what isn't (yet) used.

cavis commented on app/services/story_query_builder.rb:
```diff
@@ -0,0 +1,192 @@
+    end
+  end
+
+  def authz_filter
```
Guessing that a lot of these methods will eventually be shared with other types of query too.  (all the other models where you `include Searchable`)

cavis commented on bin/application:
```diff
@@ -29,6 +29,20 @@ WaitForMysql () {
   done
 }
 
+WaitForElasticsearch () {
```
I see how this makes sense for the tests.  But I wonder if ES-availability is really necessary for the CMS app to run at all.  (I'm always hesitant to add app-startup dependencies)

cavis replied:
Even in dev, I can see it being handy to be able to start CMS without ES.  The search endpoints just won't work.

cavis commented on config/routes.rb:
```diff
@@ -16,6 +16,7 @@
 
       resources :stories, except: [:new, :edit] do
         get 'random', on: :collection
+        get 'search', on: :collection
```
Hrmm... yeah, remind me what this is for, if we're just overriding the `?text` filter.

cavis replied:
+1 on separate endpoint, at least to start.

cavis commented on docker-compose.yml:
```diff
@@ -57,3 +59,18 @@ pg:
+  environment:
+    - cluster.name=docker-cluster
+    - bootstrap.memory_lock=true
+    - "ES_JAVA_OPTS=-Xms512m -Xmx512m"
```
Hmm... maybe these other ENVs are no longer necessary, if this changes to single-node mode.  (And should change this one to match the docker-compose-ci)

pkarman replied:
yeah I think I cargo-culted most of the env stuff from an ES example.

pkarman commented on app/controllers/api/stories_controller.rb:
```diff
@@ -94,7 +99,9 @@ def resources_base
   def filtered(resources)
     resources = resources.v4 if filters.v4?
-    resources = resources.match_text(filters.text) if filters.text?
+    Rails.logger.debug("before: #{resources}")
```
no. I'll zap.

pkarman commented:
@cavis I commented and reverted as much as I could find, and added tests for the new ES functionality.

Since this PR just implements the basic infrastructure and I didn't want it to get too big, I created a number of other tickets to describe what I think are the TODOs. See #444 

I will try adding some unit tests for the DSL query builder. We can wait on those for this PR, or merge in a subsequent PR.

cavis commented on bin/application:
```diff
@@ -70,6 +85,7 @@ ApplicationRun () {
 
   UpdateEnvironment
   WaitForMysql
+  WaitForElasticsearch
```
(I think maybe you meant to take this out)
